satpy/cf/coords.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""Set CF-compliant spatial and temporal coordinates."""
+
+import logging
+import warnings
+from collections import defaultdict
+from contextlib import suppress
+
+import numpy as np
+import xarray as xr
+from dask.base import tokenize
+from pyresample.geometry import AreaDefinition, SwathDefinition
+
+logger = logging.getLogger(__name__)
+
+
+def add_xy_coords_attrs(dataarray):
+    """Add relevant attributes to x, y coordinates."""
+    # If there are no coords, return dataarray
+    if not dataarray.coords.keys() & {"x", "y", "crs"}:
+        return dataarray
+    # If projected area
+    if _is_projected(dataarray):
+        dataarray = _add_xy_projected_coords_attrs(dataarray)
+    else:
+        dataarray = _add_xy_geographic_coords_attrs(dataarray)
+    if "crs" in dataarray.coords:
+        dataarray = dataarray.drop_vars("crs")
+    return dataarray
+
+
+def _is_projected(dataarray):
+    """Guess whether data are projected or not."""
+    crs = _try_to_get_crs(dataarray)
+    if crs:
+        return crs.is_projected
+    units = _try_get_units_from_coords(dataarray)
+    if units:
+        if units.endswith("m"):
+            return True
+        if units.startswith("degrees"):
+            return False
+    logger.warning("Failed to tell if data are projected. Assuming yes.")
+    return True
+
+
+def _try_to_get_crs(dataarray):
+    """Try to get a CRS from attributes."""
+    if "area" in dataarray.attrs:
+        if isinstance(dataarray.attrs["area"], AreaDefinition):
+            return dataarray.attrs["area"].crs
+        if not isinstance(dataarray.attrs["area"], SwathDefinition):
+            logger.warning(
+                f"Could not tell CRS from area of type {type(dataarray.attrs['area']).__name__:s}. "
+                "Assuming projected CRS.")
+    if "crs" in dataarray.coords:
+        return dataarray.coords["crs"].item()
```

Comment:
I could be wrong, but I think this logic could all be inverted or combined to make this a lot shorter. Plus, I think `SwathDefinition`s now have a `.crs` attribute so it would be good to try to use it.

```python
try:
    return dataarray.attrs["area"].crs
except (KeyError, AttributeError):
    try:
        return dataarray.coords["crs"].item()
    except (KeyError, AttributeError):
        # warning
```

Eh, I'm not sure this is much better.

Reply:
Right now, if  area is a `SwathDefinition` (and there is no `'crs'` coordinate), nothing is returned !  Do we want to change this behaviour and return the pyproj crs?  
I would suggest to do that in a separate PR, and do not change behaviour in this one @djhoese !

Reply:
Ah so my suggested change would get a CRS it previously didn't and likely add a grid_mapping for swath-based data when it previously didn't?

Sounds good to me to leave this for another time and when someone wants it. I do think all of this CF code needs to be cleaned up but it obviously isn't as easy as it looks.

Reply:
Yes exactly. And if it adds the `grid_mapping`, I fear that satpy it fails to read the saved dataset ... or it will assume it's an AreaDef in `load_cf` ... to check later on